### ISSUE TITLE
JsonString

### ISSUE THREAD

**Issue by swankjesse** [square/moshi#1232]
I was looking at a PR that hooks up `@JsonString` as an annotation for JSON literal value.

What if we add a type for a JSON literal value? A sketch of the potential API:

```
public final class JsonString {
  /** Decodes this to a Map, List, String, Double, Boolean, or null. */
  public Object toJsonValue();

  /**
   * Returns a JsonString that is equal to this. All equal JsonStrings canonicalize to
   * a value with identical toString().
   *
   * The following transformations are performed:
   * <ul>
   *   <li>Whitespace is stripped, except when inside strings
   *  <li>Fields are sorted. Note that this is a case-sensitive sorting, so `Z` precedes `a`.
   *  <li>Numeric types are encoded using Double.toString().
   * </ul>
   */
  public JsonString canonicalize();

  /** 
   * Returns true if this JSON string and {@code other} decode to equal
   * values. The following are insignificant in this comparison:
   * <ul>
   *    <li>Whitespace, except when inside strings
   *    <li>The order of fields within an object
   *    <li>String representation of numeric types. Numeric values are converted
   *          to doubles for comparison.
   * </ul>
   * 
   * Note that null is considered to be distinct from absent in this comparison.
   */
  public boolean equals(Object other);
  public int hashCode();

  /** Returns the original string encoding of this value. */
  public String toString();

  public static JsonString get(ByteString utf8Bytes);
  public static JsonString get(String string);
  public static JsonString fromJsonValue(Object object);
}
```

Plus helper methods elsewhere:

```
public class JsonReader {
  ...
  public JsonString readJsonString();
  public static JsonReader of(JsonString string);
}
```

```
public class JsonWriter {
  ...
  public JsonWriter value(JsonString string);
}
```

```
public class JsonAdapter<T> {
  ...
  public final @Nullable T fromJsonString(@Nullable Object value) {
  public final @Nullable JsonString toJsonValue(@Nullable T value) {
}
```

**Comment by swankjesse** [comment]
(And of course, you could use a `@JsonString` as a member of a field in any object, and Moshi could encode/decode it very efficiently.)

**Comment by JakeWharton** [comment]
I'm lukewarm on the idea, especially since Kotlin has inline classes and Java X+N will also have them which better serves these newtype-style wrappers. ByteString does something that doesn't exist in the JDK, but I think String does all we need here. We don't expose the serialized form backing an `any` in Wire as being a `ProtoBytes`.

**Comment by swankjesse** [comment]
I don't expect this to be implemented as just a wrapped string or ByteString. In particular, to do a good equals function we'd memoize the canonicalized string and compare those.

**Comment by JakeWharton** [comment]
Sure, but the same is true of doing equals or hash code of a proto's bytes. I'm more asking whether or not there are use cases to motivate such an abstraction and why are they here and not in Wire too? I'm skeptical as to whether Moshi needs to expose it.

**Comment by swankjesse** [comment]
I've got a bunch of code throughout various projects typed String but holding JSON. It feels good to use a specific type for that. But it's certainly not necessary.

**Comment by JakeWharton** [comment]
I guess I'm just hung up on why this deserves a named type but the library advocates using qualifier annotations for giving special meaning to a carrier type? Or why does this deserve a named type but we chose to use map/list/string/double for the document API which also represents JSON values and would benefit from type safety?

**Comment by swankjesse** [comment]
> qualifier annotations

I think qualifier annotations are best suited to cases where the Java/Kotlin object is the same, but the JSON representation is different:
 * Date formats
 * Values wrapped in unnecessary arrays
 * yes/no booleans
Stuff like this. We haven't needed these much in the applications I've written! But they're very handy for working with bad schemas.

> Map/List/string/double

These types are Java/Kotlin’s closest representations of the corresponding JSON model. A list in Java and a list in JSON are semantically equivalent!

But the semantic richness of these types is also very weak! An Instant is better than a String and a ShippingAddress is better than a Map, which is why we’ve made such a deep investment in databinding. When we have stronger types available, we prefer to use those types in the Java and Kotlin code.

In JSON we use string for lots of things:
 - customer entered text
 - dates
 - enum constants
 - identifiers

In Java, we use different types for these things. Even though it's usually fine to pass a string as an identifier, I like to use a type to keep things straight:

val smsNumber: String
val smsNumber: SmsNumber

The type also offers a natural facility to enforce constraints and add features.

I claim JsonString is a useful type because it prevents you from mixing up a JSON document containing a string `"9\" nails"` with its decoded contents.

It is also potentially useful when testing JSON: users can assert the JsonString of expected and actual outputs are equal without being needlessly sensitive to whitespace differences.

**Comment by JakeWharton** [comment]
I totally understand all of these things. I guess my hesitation is why suddenly is this the thing that needs a semantic type when we haven't done it for anything else in the many libraries we work on.

OkHttp has header value stanzas separated by `;` and with key/value pairs separated `=` for which things like order don't matter and values overwrite or combine in a subtle yet predictable way.

Proto is bytes but whose messages when opaque still have key/value pairs whose order is not significant and where encoding details like packed vs. unpacked do not affect the equality or hashCode of the encoded representation.

Even in the examples you list I don't really see a significant difference between an ISO 8601 string and a normal string as compared to a JSON string and a normal string. It's just characters that conform to an RFC which adds semantic information with regards to interpretation. I can ask an `Iso8601String` whether it takes place before or after today. I can ask a `JsonString` whether it contains the `"foo"` key. I cannot ask either of things of an opaque `String`.

But now one is relegated to a qualifier annotation whereas the other is elevated to a concrete type. I'm not fundamentally opposed to the type on principle, but has something changed here? Would you build `ProtoBytes` now in Wire to support order-indepent serialized bytes having correct equals/hashCode too?

**Comment by swankjesse** [comment]
Yeah I agree there’s no hard line on when to introduce a semantic type.

For OkHttp we have structured headers for cache-control, challenges, cookies, and media type. Adding more is a trade-off between size and utility.

For proto bytes? We could add a type for unknown fields and that would be okay, but I'm not particularly excited by it. I think the thing that makes it more interesting for JSON is avoiding decoding, either because you don't know the structure or don't need to know it.

I've written some REST APIs that accept a JSON attachment, not unlike OkHttp’s tags. From the perspective of my API I don't know how to interpret this thing and don't really want to. The best way to do this in Moshi is with a field of type Any; I claim JsonString is a bit nicer.

**Comment by somayaj** [comment]
Can this be worked on?

**Comment by ygnessin** [comment]
@somayaj this is available right now if you borrow from [this example](https://github.com/square/moshi/pull/1230) along with the latest version of Moshi
